corehq/apps/auditcare/tests/test_tasks.py

```diff
@@ -0,0 +1,97 @@
+                ip_address='10.1.2.3',
+                request_path='/a/delmar/phone/restore/?version=2.0&since=...',
+                session_key='abc123',
+                status_code=200,
+                user_agent='okhttp/4.4.1',
+                view_kwargs={'domain': 'delmar'},
+                view='corehq.apps.ota.views.restore',
+            ),
+            save_couch_doc(
+                "AccessAudit",
+                username,
+                event_date=datetime(2021, 2, 1, 3).strftime("%Y-%m-%dT%H:%M:%SZ"),
+                access_type='login',
+                description='Login Success',
+                failures_since_start=None,
+                get_data=[],
+                http_accept='text/html',
+                ip_address='10.1.3.2',
+                path_info='/a/delmar/login/',
+                post_data=[],
+                session_key='abc123',
+                user_agent='Mozilla/5.0',
+            ),
+            save_couch_doc(
+                "NavigationEventAudit",
+                username,
+                event_date=datetime(2021, 2, 1, 5).strftime("%Y-%m-%dT%H:%M:%SZ"),
+                description='User Name',
+                extra={},
+                headers={
+                    'REQUEST_METHOD': 'GET',
+                },
+                ip_address='10.1.2.3',
+                request_path='/a/sandwich/phone/restore/?version=2.0&since=...',
+                session_key='abc123',
+                status_code=200,
+                user_agent='okhttp/4.4.1',
+                view_kwargs={'domain': 'sandwich'},
+                view='corehq.apps.ota.views.restore',
+            ),
+        ]
+
+    @classmethod
+    def tearDownClass(cls):
+        delete_couch_docs(cls.couch_ids)
+        super().tearDownClass()
+
+    def tearDown(self):
+        AccessAudit.objects.all().delete()
+        NavigationEventAudit.objects.all().delete()
+        super().tearDown()
+
+    def test_copy(self):
+        NavigationEventAudit(event_date=datetime(2021, 2, 1, 4), path="just/a/checkpoint").save()
+        copy_events_to_sql()
+
+        self.assertEqual(NavigationEventAudit.objects.count(), 2)
+        self.assertEqual(
+            [e.path for e in NavigationEventAudit.objects.order_by("-event_date").all()],
+            ["just/a/checkpoint", "/a/delmar/phone/restore/?version=2.0&since=..."]
+        )
+
+        self.assertEqual(AccessAudit.objects.count(), 1)
+        self.assertEqual(AccessAudit.objects.first().path, "/a/delmar/login/")
+
+    def test_limit(self):
+        NavigationEventAudit(event_date=datetime(2021, 2, 1, 4), path="just/a/checkpoint").save()
+        copy_events_to_sql(limit=1)
+        self.assertEqual(NavigationEventAudit.objects.count(), 1)
+        self.assertEqual(NavigationEventAudit.objects.first().path, "just/a/checkpoint")
+
+        self.assertEqual(AccessAudit.objects.count(), 1)
+        self.assertEqual(AccessAudit.objects.first().path, "/a/delmar/login/")
```

Comment:
Would be nice to have tests for
- events with duplicate dates (all events should be copied exactly once, no duplicates)
  - multiple `NavigationAuditEvent` records with same `event_date`
  - multiple `AccessAudit` records with same `event_date`
  - `NavigationAuditEvent` and `AccessAudit` record with same `event_date`
- failure in middle of a copy operation, next copy operation should not produce duplicates
- back-to-back copy tasks should not produce duplicates (of the first/last record)

Reply:
> events with duplicate dates (all events should be copied exactly once, no duplicates)

I had thought there should be one sql object for each couch document. Are you saying there may be a date that has multiple couch documents but that should only be stored once in sql? This would also affect the `couch_id` logic I added, which assumes that documents are unique.

> failure in middle of a copy operation, next copy operation should not produce duplicates

Is this still relevant if I don't add bulk saving?

Reply:
Added a test that re-runs the copy in bb6a2634fe2, holding off on the others for the moment.

Reply:
You are correct, there should be exactly one SQL row per Couch document, but `event_date` is not necessarily unique. I'm concerned that if you run the script twice for the same/overlapping date range (esp. in case of errors) that it will end up processing some docs twice. A given doc should be created in SQL exactly once, even if it is processed a second time.